craft_parts/plugins/uv_plugin.py

```diff
@@ -0,0 +1,131 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2020-2024 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The uv plugin."""
+
+import shlex
+import shutil
+from typing import Literal, cast
+
+import pydantic
+from overrides import override
+
+from craft_parts import errors
+from craft_parts.plugins import validator
+
+from .base import BasePythonPlugin
+from .properties import PluginProperties
+
+
+class UvPluginProperties(PluginProperties, frozen=True):
+    """The part properties used by the uv plugin."""
+
+    plugin: Literal["uv"] = "uv"
+
+    uv_extras: set[str] = pydantic.Field(
+        default_factory=set,
+        title="Optional dependency groups",
+        description="Optional dependency groups to include when installing.",
+    )
+
+    # part properties required by the plugin
+    source: str  # pyright: ignore[reportGeneralTypeIssues]
+
+
+class UvPluginEnvironmentValidator(validator.PluginEnvironmentValidator):
+    """Check the execution environment for the uv plugin.
+
+    :param part_name: The part whose build environment is being validated.
+    :param env: A string containing the build step environment setup.
+    """
+
+    _options: UvPluginProperties
+
+    @override
+    def validate_environment(
+        self, *, part_dependencies: list[str] | None = None
+    ) -> None:
+        """Ensure the environment has the dependencies to build uv applications.
+
+        :param part_dependencies: A list of the parts this part depends on.
+        """
+        version = self.validate_dependency(
+            dependency="uv",
+            plugin_name=self._options.plugin,
+            part_dependencies=part_dependencies,
+            argument="version",
+        )
+        if not version.startswith("uv") and (
+            part_dependencies is None or "uv-deps" not in part_dependencies
+        ):
+            raise errors.PluginEnvironmentValidationError(
+                part_name=self._part_name,
+                reason=f"invalid uv version {version!r}",
+            )
+
+
+class UvPlugin(BasePythonPlugin):
+    """A plugin to build python parts."""
+
+    properties_class = UvPluginProperties
+    validator_class = UvPluginEnvironmentValidator
+    _options: UvPluginProperties
+
+    def _get_uv(self) -> str:
+        uv = shutil.which("uv")
+        # Guaranteed not-none by the validator class, which already asserted uv
```

Comment:
This is actually not guaranteed: in the (unlikely) event that the project has a part called `uv-deps` that does *not* actually provide `uv`, this would fail.
Therefore I think it's sensible to check for `None` here and fail

Reply:
Maybe we should even eliminate this and just use `uv` in the commands (getting from the part's path)? That's what most plugins do and we just fail there with a command not found.